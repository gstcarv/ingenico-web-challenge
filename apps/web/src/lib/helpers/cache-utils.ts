import type { HistoricalConversionParams } from "./currency-schema";

/**
 * Creates a consistent cache key for currency conversion requests
 */
export const createConversionCacheKey = (params: HistoricalConversionParams): string => {
    const { fromCurrency, toCurrency, amount, date } = params;
    const dateString = date ? date.toISOString().split("T")[0] : "latest";
    return `${fromCurrency}-${toCurrency}-${amount}-${dateString}`;
};

/**
 * Checks if two conversion parameters are equivalent
 */
export const areConversionParamsEqual = (
    params1: HistoricalConversionParams,
    params2: HistoricalConversionParams,
): boolean => {
    return createConversionCacheKey(params1) === createConversionCacheKey(params2);
};
