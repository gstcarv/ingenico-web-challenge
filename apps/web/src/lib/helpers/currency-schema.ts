import { z } from "zod";

export const currencyFormSchema = z.object({
    fromCurrency: z.string().min(1, "Please select a currency to convert from"),
    toCurrency: z.string().min(1, "Please select a currency to convert to"),
    amount: z.number().min(0.01, "Amount must be greater than 0"),
    date: z.date().optional(),
});

export type CurrencyFormData = z.infer<typeof currencyFormSchema>;
