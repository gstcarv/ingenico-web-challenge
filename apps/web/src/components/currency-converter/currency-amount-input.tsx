import { CurrencyInput } from "@ingenico-challenge/ui";
import { useQuery } from "@tanstack/react-query";
import { Controller, Control, useWatch } from "react-hook-form";
import { getAvailableCurrenciesQuery } from "../../queries/get-available-currencies";
import { CurrencyFormData } from "../../lib/helpers/currency-schema";
import { ReactNode } from "react";

interface CurrencyAmountInputProps {
    control: Control<CurrencyFormData>;
    suffix?: ReactNode;
}

export const CurrencyAmountInput = ({ control, suffix }: CurrencyAmountInputProps) => {
    const { data: availableCurrencies } = useQuery(getAvailableCurrenciesQuery());

    // Watch the fromCurrency to get the selected currency
    const fromCurrency = useWatch({
        control,
        name: "fromCurrency",
    });

    // Function to determine text size based on symbol length
    const getTextSizeClass = (symbol: string) => {
        if (!symbol) return "text-lg";

        const length = symbol.length;
        if (length <= 1) return "text-lg font-semibold";
        if (length <= 2) return "text-base font-semibold";
        if (length <= 3) return "text-sm font-medium";
        return "text-xs font-medium";
    };

    return (
        <Controller
            name="amount"
            control={control}
            render={({ field }) => {
                const selectedCurrency = availableCurrencies?.data?.[fromCurrency];
                const symbol = selectedCurrency?.symbol_native || selectedCurrency?.symbol || "";
                const textSizeClass = getTextSizeClass(symbol);

                return (
                    <CurrencyInput
                        id="amount"
                        value={field.value}
                        onChange={field.onChange}
                        className="w-full shadow-none border-gray-200 focus:border-blue-500 focus:ring-1 focus:ring-blue-500"
                        prefix={symbol && <span className={`${textSizeClass} text-gray-700`}>{symbol}</span>}
                        suffix={suffix}
                    />
                );
            }}
        />
    );
};
