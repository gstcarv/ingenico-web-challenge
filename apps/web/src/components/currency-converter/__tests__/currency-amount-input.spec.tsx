import { render, screen, renderHook } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { useForm, Control } from "react-hook-form";
import { CurrencyAmountInput } from "../currency-amount-input";
import { CurrencyFormData } from "../../../lib/helpers/currency-schema";
import currenciesMock from "../../../mocks/currencies.json";

// Mock the query
vi.mock("../../../queries/get-available-currencies", () => ({
    getAvailableCurrenciesQuery: () => ({
        queryKey: ["currencies"],
        queryFn: () => Promise.resolve(currenciesMock),
    }),
}));

// Mock react-query to return data immediately
vi.mock("@tanstack/react-query", async () => {
    const actual = await vi.importActual("@tanstack/react-query");
    return {
        ...actual,
        useQuery: () => ({
            data: currenciesMock,
            isLoading: false,
            error: null,
        }),
    };
});

const TestWrapper = ({ children }: { children: React.ReactNode }) => {
    const queryClient = new QueryClient({
        defaultOptions: {
            queries: { retry: false },
            mutations: { retry: false },
        },
    });

    return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

const TestComponent = ({ control, suffix }: { control: Control<CurrencyFormData>; suffix?: React.ReactNode }) => {
    return (
        <TestWrapper>
            <CurrencyAmountInput control={control} suffix={suffix} />
        </TestWrapper>
    );
};

describe("CurrencyAmountInput", () => {
    it("should render the component with correct structure", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());
        const { container } = render(<TestComponent control={result.current.control} />);

        expect(container.firstChild).toBeInTheDocument();
        expect(screen.getByRole("textbox")).toBeInTheDocument();
    });

    it("should render with USD currency symbol when fromCurrency is USD", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());
        result.current.setValue("fromCurrency", "USD");

        render(<TestComponent control={result.current.control} />);

        expect(screen.getByText("$")).toBeInTheDocument();
    });

    it("should render with EUR currency symbol when fromCurrency is EUR", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());
        result.current.setValue("fromCurrency", "EUR");

        render(<TestComponent control={result.current.control} />);

        expect(screen.getByText("€")).toBeInTheDocument();
    });

    it("should render without prefix when no currency is selected", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());

        render(<TestComponent control={result.current.control} />);

        expect(screen.queryByText("$")).not.toBeInTheDocument();
        expect(screen.queryByText("€")).not.toBeInTheDocument();
    });

    it("should render with suffix when provided", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());
        const suffix = <span>Test Suffix</span>;

        render(<TestComponent control={result.current.control} suffix={suffix} />);

        expect(screen.getByText("Test Suffix")).toBeInTheDocument();
    });

    it("should have correct CSS classes", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());
        render(<TestComponent control={result.current.control} />);

        const input = screen.getByRole("textbox");
        expect(input).toHaveClass("w-full");
    });

    it("should match snapshot", () => {
        const { result } = renderHook(() => useForm<CurrencyFormData>());
        const { container } = render(<TestComponent control={result.current.control} />);

        expect(container.firstChild).toMatchSnapshot();
    });
});
