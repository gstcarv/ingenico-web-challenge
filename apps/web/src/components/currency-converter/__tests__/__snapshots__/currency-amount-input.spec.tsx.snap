// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrencyAmountInput > should match snapshot 1`] = `
<div
  class="input-base h-16 bg-white border border-neutral-200 rounded-lg relative text-neutral-500 font-medium w-full shadow-none border-gray-200 focus:border-blue-500 focus:ring-1 focus:ring-blue-500"
>
  <input
    class="w-full h-full bg-transparent border-none outline-none text-neutral-900 font-medium placeholder:text-neutral-400 focus-visible:ring-0 text-lg px-4"
    data-testid="currency-input"
    id="amount"
    inputmode="numeric"
    placeholder="0.00"
    type="text"
    value=""
  />
</div>
`;
